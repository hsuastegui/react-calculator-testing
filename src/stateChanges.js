// @flow
export const add = (state: Object, props: Object): { total: number } => {
  return { total: state.total + state.amount }
};
export const substract = (state: Object, props: Object): { total: number } => {
  return { total: state.total - state.amount }
};
