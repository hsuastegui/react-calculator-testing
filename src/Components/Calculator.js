import React, { Component } from "react";
import { add, substract } from '../stateChanges';

class Calculator extends Component {
  static propTypes = {
    amount: React.PropTypes.number.isRequired
  };
  static defaultProps = {
    amount: 0
  };
  state = {
    amount: this.props.amount,
    total: 0
  };
  handleChange = event => {
    this.setState({
      amount: Number(event.target.value)
    });
  };
  add = () => {
    this.setState(add);
  };
  substract = () => {
    this.setState(substract);
  };
  render() {
    return (
      <section>
        <h1>Simple Calculator</h1>
        <input
          type="text"
          value={this.state.amount}
          onChange={this.handleChange}
        />
        <button id="add" onClick={this.add}>+</button>
        <button id="substract" onClick={this.substract}>-</button>
        <p>Total: <span id="total">{this.state.total}</span></p>
      </section>
    );
  }
}

export default Calculator;
