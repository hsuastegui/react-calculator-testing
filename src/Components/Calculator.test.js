import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Calculator from "./Calculator";
import { add, substract } from "../stateChanges";
//What to test: general => Renders, props/state => output, conditions/actions => states, actions => events, edge cases
describe("Calculator tests", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("has an input", () => {
    const component = shallow(<Calculator />);
    expect(component.find("input").length).toEqual(1);
  });
  it("adds", () => {
    expect(add({ total: 1, amount: 1 })).toHaveProperty("total", 2);
  });
  it("substracts", () => {
    expect(substract({ total: 1, amount: 1 })).toHaveProperty("total", 0);
  });
  it("allows user to increase score", () => {
    const component = shallow(<Calculator amount={1} />);
    component.find("button#add").simulate("click");
    expect(Number(component.find("#total").text())).toEqual(1);
  });
});
