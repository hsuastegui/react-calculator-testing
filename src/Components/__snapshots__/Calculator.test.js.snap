exports[`Calculator tests renders correctly 1`] = `
<section>
  <h1>
    Simple Calculator
  </h1>
  <input
    onChange={[Function]}
    type="text"
    value={0} />
  <button
    id="add"
    onClick={[Function]}>
    +
  </button>
  <button
    id="substract"
    onClick={[Function]}>
    -
  </button>
  <p>
    Total: 
    <span
      id="total">
      0
    </span>
  </p>
</section>
`;
